src/utils/astar.rs

```diff
@@ -327,171 +372,195 @@ impl<T: VertexDataProvider> Graph<T> {
+
+    pub fn build_indexed_path(
+        &self,
         from: usize,
         to: usize,
-        path: &mut Vec<Vector3<f32>>,
+        path: &mut Vec<usize>,
     ) -> Result<PathKind, PathError> {
         path.clear();
-        self.build_and_convert(from, to, |_, v| path.push(v.position))
-    }
 
-    /// Tries to build path from begin point to end point. Returns path kind:
-    ///
-    /// - Full: there are direct path from begin to end.
-    /// - Partial: there are not direct path from begin to end, but it is closest.
-    /// - Empty: no path available - in most cases indicates some error in input params.
-    ///
-    /// # Notes
-    ///
-    /// This is more or less naive implementation, it most certainly will be slower than specialized solutions.
-    pub fn build_and_convert<F>(
-        &mut self,
-        from: usize,
-        to: usize,
-        func: F,
-    ) -> Result<PathKind, PathError>
-    where
-        F: FnMut(usize, &T),
-    {
         if self.vertices.is_empty() {
             return Ok(PathKind::Partial);
         }
 
-        for vertex in self.vertices.iter_mut() {
-            vertex.clear();
-        }
-
         let end_pos = self
             .vertices
             .get(to)
             .ok_or(PathError::InvalidIndex(to))?
             .position;
 
-        // Put start vertex in open set.
-        let start = self
-            .vertices
-            .get_mut(from)
-            .ok_or(PathError::InvalidIndex(from))?;
-        start.state = PathVertexState::Open;
-        start.g_score = 0.0;
-        start.f_score = heuristic(start.position, end_pos);
-
-        let mut open_set_size = 1;
-        while open_set_size > 0 {
-            let mut current_index = 0;
-            let mut lowest_f_score = f32::MAX;
-            for (i, vertex) in self.vertices.iter().enumerate() {
-                if vertex.state == PathVertexState::Open && vertex.f_score < lowest_f_score {
-                    current_index = i;
-                    lowest_f_score = vertex.f_score;
-                }
-            }
+        //returns one point if the goal is the current postion
+        if from == to {
+            path.push(to);
+            return Ok(PathKind::Full);
+        }
+
+        // keeps track of which vertices we've searched
+        let mut searched_vertices = vec![false; self.vertices.len()];
+
+        // creates heap for searching
+        let mut search_heap: BinaryHeap<PartialPath> = BinaryHeap::new();
 
-            if current_index == to {
-                self.reconstruct_path(current_index, func);
-                return Ok(PathKind::Full);
+        // creates first partial path and adds it to heap
+        search_heap.push(PartialPath::new(from));
+
+        // stores best path found
+        let mut best_path = PartialPath::default();
+
+        // search loop
+        // TODO: don't hard code max search iterations
+        for _ in 0..1000 {
+            // breakes loop if heap is empty
```

Comment:
when it comes to max search iterations, I don't know if we want to: 
- Hard code it. 
- Let the user handle it.
- Remove it.
They all have their benefits and drawbacks. I would make it a property of Graph, give it a default value, and let the user change it.